src/client/Client.cc

```diff
@@ -11232,6 +11236,11 @@ int64_t Client::_preadv_pwritev_locked(Fh *fh, const struct iovec *iov,
 {
     ceph_assert(ceph_mutex_is_locked_by_me(client_lock));
 
+    if (fh == NULL || !fh_exists(fh)) {
```

Comment:
Have you checked other _unchecked_ uses to the file handles?

Reply:
... and this isn't really async specific, isn't it? (given that the test case is in nonblocking.cc).

Reply:
> ... and this isn't really async specific, isn't it? (given that the test case is in nonblocking.cc).

right, not specific to async

Reply:
> Have you checked other _unchecked_ uses to the file handles?

I found most cases handled in respective APIs, this seems to be the only condition(so far) that wasn't checked. If you can think of any/find any do let me know.

Reply:
is it possible that a file handle has no inode associated with it?

Reply:
> is it possible that a file handle has no inode associated with it?

nope ... a file handle is always created *for* an inode (after the dentry is _created_ or _opened_)

Reply:
> > is it possible that a file handle has no inode associated with it?
> 
> nope ... a file handle is always created _for_ an inode (after the dentry is _created_ or _opened_)

is there any chance that the inode reference file handle is holding is lost? i dont know if this even makes sense but just curious

Reply:
> > > is it possible that a file handle has no inode associated with it?
> > 
> > 
> > nope ... a file handle is always created _for_ an inode (after the dentry is _created_ or _opened_)
> 
> is there any chance that the inode reference file handle is holding is lost? i dont know if this even makes sense but just curious

yeah, its possible ... maybe via some bad code or overlooked cases in different part of the code ?
so far it doesn't seem so

Reply:
> > ... and this isn't really async specific, isn't it? (given that the test case is in nonblocking.cc).
> 
> right, not specific to async

Move the test to generic location then?

Reply:
can goto `src/test/libcephfs/test.cc`

Reply:
We have two kind of APIs:
1. that accept `fd` (an int) as an arg
2. that accept pointer to file handle as an arg. (sync + async)

Most APIs fall in 1st case and all of them have a check for file handle pointer's validness:
```
Fh *fh = get_filehandle(fd);
    if (!fh)
      return -CEPHFS_EBADF;
```

we already have tests for this in `test.cc` (test case named `BadFileDesc`). So the sync calls are sorted.

It is only the 2nd case where APIs directly make use of the file handle pointers and there are only three of them i.e. `ll_preadv_pwritev`, `ll_writev` and `ll_readv`. 

My current implementation adds the check to helper func `_preadv_pwritev_locked` that all three of these use but the APIs that fall in 1st case make sure they check for it before passing in the `fh` to the helpers. So my question is what should be done, should I add these checks in all the three APIs (which i feel looks redundant) or keep it the way it is i.e. have the helper check for it? @mchangir @vshankar

Reply:
> My current implementation adds the check to helper func `_preadv_pwritev_locked` that all three of these use but the APIs that fall in 1st case make sure they check for it before passing in the `fh` to the helpers. So my question is what should be done, should I add these checks in all the three APIs (which i feel looks redundant) or keep it the way it is i.e. have the helper check for it? @mchangir @vshankar

Just have the fd checked in `_preadv_pwritev_locked`.

Reply:
this breaks up higher level APIs, eventually every API calls `_preadv_pwritev_locked` but the thing is is high level APIs add `fh` to the fd_map while low level APIs make use of `ll_unclosed_fh_set` so if the file is opened using high level API like open or openat then those calls will fails since it wont be able to find the `fh` in `ll_unclosed_fh_set` as it was never added to it